pkg/cri/resource-manager/control/blockio/blockio.go

```diff
@@ -133,25 +164,282 @@ func (ctl *blockio) BlockIOClass(c cache.Container) string {
+	devMatches := map[string]string{} // {devNodeOrSymlink: origin}
+	for _, devWildcard := range devWildcards {
+		devWildcardMatches, err := filepath.Glob(devWildcard)
+		if err != nil {
+			errors = append(errors, fmt.Errorf("bad device wildcard %#v: %+v", devWildcard, err))
+			continue
+		}
+		if len(devWildcardMatches) == 0 {
+			log.Warn("deviceParametersToOci: device wildcard %#v does not match any device nodes", devWildcard)
+			continue
+		}
+		for _, devMatch := range devWildcardMatches {
+			if devMatch != devWildcard {
+				origin = fmt.Sprintf("from wildcard %#v", devWildcard)
+			} else {
+				origin = ""
+			}
+			devMatches[devMatch] = strings.TrimSpace(fmt.Sprintf("%v %v", devMatches[devMatch], origin))
+		}
+	}
+	if len(devMatches) == 0 {
+		log.Warn("no device node matches for any of Devices %#v, parameters ignored", devWildcards)
+		return blockDevices, nil
+	}
+
+	// 2. Find out real device nodes behind symlinks
+	// Example: devRealPaths["/dev/sda"] == "from symlink \"/dev/disk/by-id/ata-VendorSSD\""
+	devRealpaths := map[string]string{} // {devNode: origin}
+	for devMatch, devOrigin := range devMatches {
+		realDevNode, err := filepath.EvalSymlinks(devMatch)
+		if err != nil {
+			log.Warn("cannot filepath.EvalSymlinks(%#v): %+v", devMatch, err)
+			continue
+		}
+		if realDevNode != devMatch {
+			origin = fmt.Sprintf("from symlink %#v %v", devMatch, devOrigin)
+		} else {
+			origin = devOrigin
+		}
+		devRealpaths[realDevNode] = strings.TrimSpace(fmt.Sprintf("%v %v", devRealpaths[realDevNode], origin))
+	}
+
+	// 3. Filter out everything but block devices that are not partitions
+	// Example: blockDevices["/dev/sda"] == {Major: 8, Minor: 0, Origin: "..."}
+	for devRealpath, devOrigin := range devRealpaths {
+		origin := ""
+		if devOrigin != "" {
+			origin = fmt.Sprintf(" (origin: %s)", devOrigin)
+		}
+		fileInfo, err := os.Stat(devRealpath)
+		if err != nil {
+			log.Warn("cannot os.Stat(%#v): %+v%s", devRealpath, err, origin)
+			continue
+		}
+		fileMode := fileInfo.Mode()
+		if fileMode&os.ModeDevice == 0 {
+			errors = append(errors, fmt.Errorf("file %#v is not a device%s", devRealpath, origin))
+			continue
+		}
+		if fileMode&os.ModeCharDevice != 0 {
+			log.Error("file %#v is a character device%s", devRealpath, origin)
+			continue
+		}
+		sys, ok := fileInfo.Sys().(*syscall.Stat_t)
+		major := unix.Major(sys.Rdev)
+		minor := unix.Minor(sys.Rdev)
+		if !ok {
+			log.Error("cannot get syscall stat_t from %#v: %+v%s", devRealpath, err, origin)
+			continue
+		}
+		if minor&0xf != 0 {
+			log.Warn("skipping %#v: cannot weight/throttle partitions%s", devRealpath, origin)
+			continue
+		}
+		blockDevices[devRealpath] = BlockDeviceInfo{
+			Major:  int64(major),
+			Minor:  int64(minor),
+			Origin: strings.TrimSpace(fmt.Sprintf("%v %v", blockDevices[devRealpath].Origin, devOrigin)),
+		}
+	}
+	return blockDevices, nil
```

Comment:
Was there supposed to be an error Join/test in the end before that `return` like above:
```
    err = utils.JoinErrors(errors)
￼    if err != nil {
        return nil, err
￼    }
    return blockDevices, nil
```

Reply:
Thanks! Rewritten so that all parameter sets are handled even in case of error.